Lighten Docker Compose setup

pacodelaluna opened the pull request:
#### What? Why?

- Relates to #8823

While working on OFN lately, I have found a way to lighten the actual Docker setup by launching Webpack and Sidekiq inside the current container, thanks to a new dedicated Procfile.


#### What should we test?
- Delete the container related to webpack and worker (3gb each)
- Build Docker Compose config
- Check that no new container were created

#### Release notes
- Lighten Docker Compose setup

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [x] Technical changes only
- [ ] Feature toggled

rioug commented on compose.yaml:
```diff
@@ -27,51 +28,19 @@ services:
       - redis
-      - webpack
     environment:
+      DOCKER: true
+      CHROME_URL: http://browserless:4000
```
What are these used for ? Are you trying to use https://github.com/browserless/browserless  to run system spec on docker ?

rioug reviewed (commented):
Nice one ! As a next step, I think we could further trim down the Docker image by using an official ruby image instead of using Ubuntu. What do you think ?

rioug commented:
I managed to run the ofn with this docker setup, but it doesn't change anything for system spec. It still not possible to run them in docker.

mkllnk commented on app/models/enterprise.rb:
```diff
@@ -127,6 +127,9 @@ class Enterprise < ApplicationRecord
               message: Spree.t('errors.messages.invalid_instagram_url')
             }, allow_blank: true
   validate :validate_white_label_logo_link
+  validates :external_billing_id,
```
I think this commit belongs to another pull request.

pacodelaluna replied:
Yep, sorry about that, just fix it.

mkllnk reviewed (commented):
Nice work, thank you. There's just one additional commit, otherwise I approve.

dacook commented:
> Nice one ! As a next step, I think we could further trim down the Docker image by using an official ruby image instead of using Ubuntu. What do you think ?

Wasn't that the point of the [Alpine image](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/alpine.Dockerfile)?
 
From what I understand, Alpine looks like a good option because it requires less resources to run on top of Windows or Mac.

rioug commented:
> Wasn't that the point of the [Alpine image](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/alpine.Dockerfile)?

Indeed I had forgotten about this, maybe we should try to be using this image by default ?

mkllnk commented:
> maybe we should try to be using this image by default ?

Okay. I like the idea of saving resources.

pacodelaluna commented:
Good point, I was actually missing this new Dockerfile! I have just added to the Compose configuration, and it is working really good.
As suggested by @rioug, I have removed the former Dockerfile and rename the alpine one with the default name.
The system tests are still not running in it, but it can be a next topic.
Please tell me if any issue.

rioug reviewed (approved):
Nice one ! Thanks for your help :pray:

rioug commented:
> The system tests are still not running in it, but it can be a next topic.

If that's something you want to look into, checkout : https://evilmartians.com/chronicles/system-of-a-test-setting-up-end-to-end-rails-testing . I tried that a while ago but unfortunately I did not get anywhere.

dacook reviewed (changes requested):
Just to clarify: The webpack, sidekiq and rails server containers all used the same dockerfile, therefore created three duplicate containers. This brings them all together.

Also, you've changed it to now use the Alpine base image instead of Ubuntu 20. This is to use less resources.

The Ubuntu 20 dockerfile has been deleted. Maybe we should keep this, at least in the short term as a reference? Maybe some people are still using it. 

In any case, it would be nice to have some release notes that can be posted in #dev to announce the change and describe how to clean up the old ubuntu image if you're no longer using it to save disk space.

Also, the readme needs updating to reflect this: https://github.com/openfoodfoundation/openfoodnetwork/blob/master/docker/README.md#L42:~:text=Choose%20and%20Build%20Docker%20Image%3A

dacook commented:
I just tested this (Alpine), starting with no docker images or volumes.

It used approx 5gb hard disk, which seems not bad. Still took a while to fully set up (at least 10 mins I think), but most of that was time installing gems and node modules.
<img width="619" alt="Screenshot 2025-02-04 at 1 14 46 pm" src="https://github.com/user-attachments/assets/a5540fad-130c-4196-bae6-701547b0595b" />

I was able to run the web server and specs. The precompile message keeps coming up, but I'm sure that's an existing issue.
![Screenshot 2025-02-04 at 1 23 16 pm](https://github.com/user-attachments/assets/f6da326e-e6e8-4dfd-a2c5-3d0d39b49025)
Out of curiosity I tried browser specs. It seems to just hang..
![Screenshot 2025-02-04 at 1 41 02 pm](https://github.com/user-attachments/assets/67ae2d89-ff09-4967-8f5d-f680240e7956)

dacook commented:
`±` Just confirming we still need to (in my opinion):
* update the docker/readme.md
* retain the ubuntu docker file for reference (eg as Ubuntu.dockerfile)

pacodelaluna commented:
I have just added the suggestions for the Docker configuration. I have actually missed the Docker scripts part, I am not using them as I don't know exactly how to use them. But I totally agree that it is better to avoid friction in case other devs are using them.

About a message to the Slack #dev channel, I was thinking of something like:

Docker local configuration update:
The Docker Compose configuration is now more efficient thanks to the `Profile.docker` setup. It is allowing us to launch the Rails app, the Webpack server and the Sidekiq proc in the same container, instead of running 3 different containers to do so.
Also, we have just changed the default Dockerfile image to be the Alpine one from now. It makes more sense for development. We are keeping the old Ubuntu image for support in case some of you are using it, it is now named `Dockerfile.ubuntu`.

Please let me know if it needs to be adjusted.

dacook reviewed (approved):
Great, looks good to me!

dacook commented:
@pacodelaluna would you like to do the honours and post in #dev?

dacook commented:
I went ahead and posted ☑️
